vowpalwabbit/learner.cc

```diff
@@ -36,207 +30,251 @@ const char* to_string(prediction_type_t prediction_type)
+        _context.template process<example, learn_ex>(*ec);
+      else if (ec->end_pass)
+        _context.template process<example, end_pass>(*ec);
+      else if (is_save_cmd(ec))
+        _context.template process<example, save>(*ec);
+      else
+        _context.template process<example, learn_ex>(*ec);
+    }
+  private:
+    context_type _context;
+  };
+  
+  template<typename context_type>
+  class multi_example_handler {
+  private:
+    bool complete_multi_ex(example* ec)
     {
-      cout << "Something is wrong---an example with no choice.  Do you have all 0 features? Or multiple empty lines?"
-           << endl;
+      auto& master = _context.get_master();
+      const bool is_test_ec = master.p->lp.test_label(&ec->l);
+      const bool is_newline = (example_is_newline_not_header(*ec) && is_test_ec);
+      if (!is_newline) {
+        ec_seq.push_back(ec);
+      } else {
+        VW::finish_example(master, *ec);
+      }
+      return is_newline;
+    }
+    
+    bool try_complete_multi_ex(example* ec) {
+      if (ec->indices.size() > 1)  // 1+ nonconstant feature. (most common case first)
+        return complete_multi_ex(ec);
+      else if (ec->end_pass)
+        _context.template process<example, end_pass>(*ec);
+      else if (is_save_cmd(ec))
+        _context.template process<example, save>(*ec);
+      else
+        return complete_multi_ex(ec);
       return false;
     }
-    return true;  // example complete
-  }
-  ec_seq.push_back(ec);
-  return false;
-}
-
-template <void (*f)(vw&, multi_ex&)>
-void dispatch_multi_ex(vw& all, multi_ex& ec_seq)
-{
-  f(all, ec_seq);                   // call learn or predict
-  VW::finish_example(all, ec_seq);  // clean up
-}
-
-template <void (*f)(vw&, multi_ex&)>
-void dispatch_multi_ex(vw& all, example* ec, multi_ex& ec_seq)
-{
-  if (complete_multi_ex(ec, ec_seq, all))
+    
+  public:
+    multi_example_handler(const context_type context)
+    : _context(context) {}
+    
+    ~multi_example_handler() {
+      if (!ec_seq.empty()) {
+        _context.template process<multi_ex, learn_multi_ex>(ec_seq);
+      }
+    }
+    
+    void on_example(example* ec) {
+      if (try_complete_multi_ex(ec)) {
+        _context.template process<multi_ex, learn_multi_ex>(ec_seq);
+        ec_seq.clear();
+      }
+    }
+    
+  private:
+    context_type _context;
+    multi_ex ec_seq;
+  };
+  
+  class ready_examples_queue {
+  public:
+    ready_examples_queue(vw& master)
+    : _master(master) {}
+    
+    example* pop() {
+      return !_master.early_terminate ? VW::get_example(_master.p) : nullptr;
+    }
+  private:
+    vw& _master;
+  };
+  
+  class custom_examples_queue {
+  public:
+    custom_examples_queue(v_array<example*> examples)
+    : _examples(examples) {}
+    
+    example* pop() {
+      return _index < _examples.size() ? _examples[_index++] : nullptr;
+    }
+  private:
+    v_array<example*> _examples;
```

Comment:
Can this be a std::deque or std::list?

Reply:
It technically can, but the only reason for this class to exist - to use it as adapter between driver and parse_dispatch_loop.h:parse_dispatch. So, it is easier to have the same container here as we fill in par_dispatch. It can be replaced there, but in this case diff is going to expand further (at least to all parser interfaces), so it is probably better to do in separate pr.